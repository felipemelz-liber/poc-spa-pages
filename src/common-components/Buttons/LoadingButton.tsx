import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { SuccessButton, LoadingSpinner } from 'liber-components';
import { LoadingContainer } from './LoadingButton.styles';

const LoadingButton = ({
  ButtonComponent,
  children,
  loading,
  onClick,
  disabled,
  size,
  spinnerColor,
  ...rest
}) => {
  const buttonRef = useRef(null);
  const [buttonWidth, setButtonWidth] = useState(0);

  useEffect(() => {
    setButtonWidth(buttonRef.current ? buttonRef.current.offsetWidth : 0);
  }, [buttonRef.current]);

  const getSpinnerSize = () => (size === 'small' ? '14px' : '22px');
  return (
    <ButtonComponent
      ref={buttonRef}
      size={size}
      version={2}
      onClick={onClick}
      disabled={disabled || loading}
      {...rest}
    >
      {loading ? (
        <LoadingContainer minWidth={buttonWidth - 32}>
          <LoadingSpinner color={spinnerColor} size={getSpinnerSize()} />
        </LoadingContainer>
      ) : (
        children
      )}
    </ButtonComponent>
  );
};

LoadingButton.propTypes = {
  ButtonComponent: PropTypes.elementType,
  children: PropTypes.node,
  loading: PropTypes.bool,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  spinnerColor: PropTypes.string,
};

LoadingButton.defaultProps = {
  ButtonComponent: SuccessButton,
  children: 'BotÃ£o',
  loading: false,
  onClick: () => null,
  disabled: false,
  size: 'medium',
  spinnerColor: 'white',
};

export default LoadingButton;
