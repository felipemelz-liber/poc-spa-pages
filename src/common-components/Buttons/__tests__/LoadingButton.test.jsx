import React from 'react';
import sinon from 'sinon';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { PrimaryButton, LoadingSpinner } from 'liber-components';
import LoadingButton from '../LoadingButton';

describe('LoadingButton component tests', () => {
  configure({ adapter: new Adapter() });

  const props = {
    ButtonComponent: PrimaryButton,
    children: 'Nome botÃ£o',
    loading: false,
    onClick: sinon.spy(),
    disabled: false,
  };
  const sizes = ['small', 'medium', 'large'];

  it('should render correctly without props', () => {
    const wrapper = shallow(<LoadingButton />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  sizes.forEach(size => {
    it(`should render correctly with all props and size ${size}`, () => {
      const wrapper = shallow(<LoadingButton {...props} size={size} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
  it('should render correctly with disabled prop', () => {
    const wrapper = shallow(<LoadingButton {...props} disabled />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly with custom spinner color', () => {
    const wrapper = shallow(<LoadingButton {...props} spinnerColor="#009dff" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly while loading', () => {
    const wrapper = shallow(<LoadingButton {...props} loading />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render spinner and be disabled when loading', () => {
    const wrapper = shallow(<LoadingButton {...props} loading />);
    const spinner = wrapper.find(LoadingSpinner);
    expect(spinner.length).toBe(1);
    expect(wrapper.prop('disabled')).toBeTruthy();
  });
  it('should call onClick when clicked', () => {
    const wrapper = shallow(<LoadingButton {...props} />);
    const { onClick } = props;
    wrapper.simulate('click');
    expect(onClick.called).toBeTruthy();
  });
});
