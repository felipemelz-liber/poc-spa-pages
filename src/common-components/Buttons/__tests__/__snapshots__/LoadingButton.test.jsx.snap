// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingButton component tests should render correctly while loading 1`] = `
<Buttons__PrimaryButton
  disabled={true}
  onClick={[Function]}
  size="medium"
  version={2}
>
  <styled.div
    minWidth={-32}
  >
    <LoadingSpinner
      color="white"
      size="22px"
      thickness={null}
    />
  </styled.div>
</Buttons__PrimaryButton>
`;

exports[`LoadingButton component tests should render correctly with all props and size large 1`] = `
<Buttons__PrimaryButton
  disabled={false}
  onClick={[Function]}
  size="large"
  version={2}
>
  Nome botão
</Buttons__PrimaryButton>
`;

exports[`LoadingButton component tests should render correctly with all props and size medium 1`] = `
<Buttons__PrimaryButton
  disabled={false}
  onClick={[Function]}
  size="medium"
  version={2}
>
  Nome botão
</Buttons__PrimaryButton>
`;

exports[`LoadingButton component tests should render correctly with all props and size small 1`] = `
<Buttons__PrimaryButton
  disabled={false}
  onClick={[Function]}
  size="small"
  version={2}
>
  Nome botão
</Buttons__PrimaryButton>
`;

exports[`LoadingButton component tests should render correctly with custom spinner color 1`] = `
<Buttons__PrimaryButton
  disabled={false}
  onClick={[Function]}
  size="medium"
  version={2}
>
  Nome botão
</Buttons__PrimaryButton>
`;

exports[`LoadingButton component tests should render correctly with disabled prop 1`] = `
<Buttons__PrimaryButton
  disabled={true}
  onClick={[Function]}
  size="medium"
  version={2}
>
  Nome botão
</Buttons__PrimaryButton>
`;

exports[`LoadingButton component tests should render correctly without props 1`] = `
<Buttons__SuccessButton
  disabled={false}
  onClick={[Function]}
  size="medium"
  version={2}
>
  Botão
</Buttons__SuccessButton>
`;
